### ISSUE TITLE
Twoslash output is inaccessible

### ISSUE THREAD

**Issue by delucis** [twoslashes/twoslash#32]
(Reposted from https://github.com/shikijs/twoslash/issues/195)

---

Hey there 👋 I noticed that currently TwoSlash produces inaccessible output.

This issue is not a comprehensive accessibility audit, but I wanted to draw attention to some key issues that would need addressing to make TwoSlash code blocks usable by more users.

The references here are to Web Content Accessibility Guidelines (WCAG) 2.2

### [Guideline 2.1 Keyboard Accessible](https://www.w3.org/WAI/WCAG22/quickref/#keyboard-accessible)

> All [functionality](https://www.w3.org/TR/WCAG22/#dfn-functionality) of the content is operable through a [keyboard interface](https://www.w3.org/TR/WCAG22/#dfn-keyboard-interface) without requiring specific timings for individual keystrokes

Twoslash tooltips are only available via mouse/pointer hover, so fails most of WCAG Guideline 2.1 regarding keyboard accessibility: it should be possible for users to access the content of tooltips using only the keyboard.

### [Guideline 4.1 Compatible](https://www.w3.org/WAI/WCAG22/quickref/#compatible)

> Maximize compatibility with current and future user agents, including assistive technologies.

Twoslash tooltips are not available to screen readers: navigating a code sample using Twoslash, there is no way for screenreaders to access the content of tooltips.

## Other details

There are some smaller things like the following, but these are kind of moot given the bigger issues:

### [1.4.13 Content on Hover or Focus](https://www.w3.org/TR/WCAG22/#content-on-hover-or-focus)

> A [mechanism](https://www.w3.org/TR/WCAG22/#dfn-mechanism) is available to dismiss the additional content without moving pointer hover or keyboard focus, unless the additional content communicates an [input error](https://www.w3.org/TR/WCAG22/#dfn-input-error) or does not obscure or replace other content;

Tooltip content is not dismissible via a mechanism other than moving the pointer’s hover position.

**Comment by antfu** [comment]
I am not very familiar with a11y myself - I wonder what would be your solutions in mind?

From my personal point of view, tooltips from Twoslash are usually "additional information" but not the main content, which serves as "good to have" but also okay to go without. I imagine myself as a keyboard user or screen reader; I would not want my focus or reader to be "trapped" inside the tons of type information when reading a code snippet. For example, something like:

```ts
const a = Math.min(1, 2)
```

It can be read as 

```ts
const a | const a: number | = Math | var Math: Math | An intrinsic object that provides basic mathematics functionality and constants. | (method) Math.min(...values: number[]): number | @param values — Numeric expressions to be evaluated. | (1, 2)
```

- if we make tooltips "accessible". In a way, I consider the experience much worse. I would rather hide some "secondary, optional information" instead of having a bad UX.

I can be very biased. So I'd love to hear what you suggest we improve.
